File: backend/core/proposal_copy.py
"""User-facing copy for preliminary structural proposals (not code-compliant design)."""

from __future__ import annotations

from schemas.site import StructuralRecommendations

PROPOSAL_DISCLAIMER = (
    "This is a preliminary model-generation proposal only. "
    "It is not a code-compliant structural design. "
    "Final member sizes, wind loads, connections and foundations must be "
    "verified by a licensed structural engineer."
)

_SECTION_SIZING_NOTE = (
    "Starting sections suggested using preliminary EC3-style capacity checks "
    "and heuristic load assumptions — not a full structural verification."
)


def format_bracing_enabled(rec: StructuralRecommendations) -> str:
    parts: list[str] = []
    if rec.x_bracing:
        parts.append("wall X-bracing")
    if rec.roof_bracing:
        parts.append("roof X-bracing")
    if rec.gable_bracing:
        parts.append("gable X-bracing")
    if rec.sag_rods:
        parts.append("anti-sag rods")
    if not parts:
        return "Bracing: none suggested."
    return f"Bracing enabled: {', '.join(parts)}."


def build_proposal_rationale(
    *,
    site_location: str,
    mean_wind_ms: float,
    exposure_proxy_ms: float,
    terrain_label: str,
    load_index: float,
    effective_load_index: float,
    rec: StructuralRecommendations,
    bay_mm: float,
    n_frames: int,
    width_m: float,
    length_m: float,
    height_m: float,
    use_case: str,
    prelim_roof_kn_m2: float,
    prelim_column_m_knm: float,
    prelim_chord_n_kn: float | None,
    bracing_profile: str = "L50x50",
) -> list[str]:
    """Structured rationale bullets with honest preliminary-design wording."""
    rationale: list[str] = []

    if site_location:
        rationale.append(f"Site: {site_location}.")
    rationale.append(
        f"Climate estimate: mean wind ~{mean_wind_ms:.1f} m/s, "
        f"exposure proxy {exposure_proxy_ms:.1f} m/s. "
        "Code wind speed was not calculated."
    )
    rationale.append(f"Selected terrain: {terrain_label}.")
    rationale.append(f"Internal Steelera load index: {load_index:.1f}.")
    if effective_load_index > load_index + 0.05:
        rationale.append(
            f"Conservative sizing floor applied: effective index "
            f"{effective_load_index:.1f} (internal minimum)."
        )

    bays = bay_mm / 1000.0
    frame_type = (
        f"{rec.truss_type.upper()} truss" if rec.use_truss else "portal frames"
    )
    rationale.append(
        f"Suggested configuration: {frame_type}, {rec.column_profile} columns, "
        f"~{bays:.2f} m bays, {n_frames} frames."
    )

    section_lines = [f"Columns: {rec.column_profile}"]
    if rec.use_truss and rec.truss_chord_profile:
        section_lines.append(f"Truss chords: {rec.truss_chord_profile}")
    if rec.use_truss and rec.truss_web_profile:
        section_lines.append(f"Truss webs: {rec.truss_web_profile}")
    section_lines.append(f"Bracing: {bracing_profile}")
    rationale.append(
        "Suggested starting sections:\n" + "\n".join(f"  - {s}" for s in section_lines)
    )

    rationale.append(format_bracing_enabled(rec))

    load_detail = (
        f"Estimated frame loads for sizing checks: roof {prelim_roof_kn_m2:.2f} kN/m², "
        f"column M ~{prelim_column_m_knm:.0f} kN·m"
    )
    if prelim_chord_n_kn is not None:
        load_detail += f", chord N ~{prelim_chord_n_kn:.0f} kN."
    else:
        load_detail += "."
    rationale.append(f"{_SECTION_SIZING_NOTE} {load_detail}")

    if use_case.strip():
        rationale.append(f"Use case: {use_case.strip()}.")
    rationale.append(
        f"Clear span {width_m:.1f} m × {length_m:.1f} m length, "
        f"{height_m:.1f} m eave."
    )
    rationale.append(PROPOSAL_DISCLAIMER)
    return rationale

File: frontend/src/lib/proposal-copy.ts
import type { SiteContext } from "@/types/site";

import type { StructuralRecommendations } from "@/types/site";

import { surroundingsLabel } from "@/lib/site-surroundings";



export const PROPOSAL_DISCLAIMER =

  "This is a preliminary model-generation proposal only. " +

  "It is not a code-compliant structural design. " +

  "Final member sizes, wind loads, connections and foundations must be verified by a licensed structural engineer.";

export const PROPOSAL_DISCLAIMER_SHORT =
  "Starting model only — verify with a structural engineer before construction.";



export function terrainLabelLong(terrain: string): string {

  const labels: Record<string, string> = {

    "0": "Coastal / open (Cat 0)",

    II: "Open terrain (Cat II)",

    III: "Suburban (Cat III)",

    IV: "Urban (Cat IV)",

  };

  return labels[terrain] ?? terrain;

}



export function formatClimateLine(site: SiteContext): string {

  return (

    `Climate estimate: mean wind ~${site.mean_wind_ms.toFixed(1)} m/s, ` +

    `exposure proxy ${site.design_wind_proxy_ms.toFixed(1)} m/s. ` +

    `Code wind speed was not calculated.`

  );

}



/** Primary terrain line for proposals — shows final selected context only. */

export function formatSiteTerrainLine(site: SiteContext): string {

  const applied = site.surroundings_applied ?? "auto";

  const terrainLabel =

    applied === "auto"

      ? `Terrain: ${terrainLabelLong(site.terrain_class)}`

      : `Selected terrain: ${terrainLabelLong(site.terrain_class)}`;

  return `${terrainLabel} · load index ${site.load_index.toFixed(1)}`;

}



/** Optional lineage when user override changed terrain from map detection. */

export function formatSiteOverrideNote(site: SiteContext): string | null {

  const applied = site.surroundings_applied ?? "auto";

  if (applied === "auto") return null;

  if (

    site.detected_terrain_class == null ||

    site.detected_load_index == null ||

    (site.detected_terrain_class === site.terrain_class &&

      Math.abs(site.detected_load_index - site.load_index) < 0.05)

  ) {

    return `User surroundings: ${surroundingsLabel(applied)}.`;

  }

  return (

    `Initial map detection: ${terrainLabelLong(site.detected_terrain_class)} · ` +

    `load index ${site.detected_load_index.toFixed(1)} · ` +

    `${surroundingsLabel(applied)} override applied.`

  );

}



export function formatSiteSummary(site: SiteContext): string {

  const overrideNote = formatSiteOverrideNote(site);

  const lines = [

    `Site data for ${site.location_label || "your location"}:`,

    `• ${formatClimateLine(site)}`,

    `• ${formatSiteTerrainLine(site)}`,

  ];

  if (overrideNote) {

    lines.push(`• ${overrideNote}`);

  }

  return lines.join("\n");

}



export function formatSuggestedConfiguration(rec: StructuralRecommendations): string {

  const bays = (rec.bay_spacing_mm / 1000).toFixed(2).replace(/\.?0+$/, "");

  const frameType = rec.use_truss

    ? `${rec.truss_type.toUpperCase()} truss`

    : "portal frames";

  return `Suggested configuration: ${frameType}, ${rec.column_profile} columns, ~${bays} m bays.`;

}



export function formatStartingSections(

  rec: StructuralRecommendations,

  bracingProfile = "L50x50",

): string {

  const lines = [`Columns: ${rec.column_profile}`];

  if (rec.use_truss && rec.truss_chord_profile) {

    lines.push(`Truss chords: ${rec.truss_chord_profile}`);

  }

  if (rec.use_truss && rec.truss_web_profile) {

    lines.push(`Truss webs: ${rec.truss_web_profile}`);

  }

  lines.push(`Bracing: ${bracingProfile}`);

  return lines.map((l) => `  - ${l}`).join("\n");

}



export function formatBracingEnabled(rec: StructuralRecommendations): string {

  const parts: string[] = [];

  if (rec.x_bracing) parts.push("wall X-bracing");

  if (rec.roof_bracing) parts.push("roof X-bracing");

  if (rec.gable_bracing) parts.push("gable X-bracing");

  if (rec.sag_rods) parts.push("anti-sag rods");

  if (parts.length === 0) return "Bracing: none suggested.";

  return `Bracing enabled: ${parts.join(", ")}.`;

}


